common/lib/xmodule/xmodule/course_module.py

```diff
@@ -269,6 +269,44 @@ def get_available_providers():
     return available_providers
 
 
+class DiscussionTopicMapping(Dict):
+    """
+    DiscussionTopicMapping field, which includes validation of the topic id
+    """
+
+    _default = {}
+
+    def from_json(self, value):
+        """
+        Return python dict type. Performs validation on id key for each discussion topic
+        """
+        if value is None:
+            return None
+        """
+        if statement checks to make sure value is a dict,
+        each topic in dict has a key id and the id value only has valid characters
+        """
+        if isinstance(value, dict):
+            for key in value:
+                if "id" in value[key].keys():
+                    if not self._validate_topic_id_value(value[key]["id"]):
+                        raise ValueError("The only special characters that are supported in the id key are underscore, hyphen, and period.")
+                else:
+                    raise ValueError("All topics need a id key")
+        else:
+            raise TypeError('Value stored in a Dict must be None or a dict, found %s' % type(value))
+        return value
+    enforce_type = from_json
+
+    def _validate_topic_id_value(self, value):
```

Comment:
A quick read of this function, I'm not actually sure what it's trying to do, especially with the lambda function. Can we add a comment or try to make it a bit more readable?